jax/_src/lax/linalg.py

```diff
@@ -1938,28 +2017,32 @@ def _householder_product_cpu_gpu_lowering(ctx, a, taus, *,
     platform='rocm')
 
 
-def _qr_impl(operand, *, full_matrices):
-  q, r = dispatch.apply_primitive(qr_p, operand, full_matrices=full_matrices)
-  return q, r
+def _qr_impl(operand, *, pivoting, full_matrices):
+  q, r, *p = dispatch.apply_primitive(qr_p, operand, pivoting=pivoting,
+                                      full_matrices=full_matrices)
+  return (q, r, p[0]) if pivoting else (q, r)
 
-def _qr_abstract_eval(operand, *, full_matrices):
+# TODO: determine if dtype updating is needed.
+def _qr_abstract_eval(operand, *, pivoting, full_matrices):
   if isinstance(operand, ShapedArray):
     if operand.ndim < 2:
       raise ValueError("Argument to QR decomposition must have ndims >= 2")
     *batch_dims, m, n = operand.shape
     k = m if full_matrices else core.min_dim(m, n)
     q = operand.update(shape=(*batch_dims, m, k))
     r = operand.update(shape=(*batch_dims, k, n))
+    p = operand.update(shape=(*batch_dims, n), dtype=np.dtype(np.int32))
   else:
     q = operand
     r = operand
-  return q, r
+    p = operand.update(dtyp=np.dtype(np.int32))
```

Comment:
This won't work in general (`update` won't be defined if the input isn't a `ShapedArray`). The standard approach (e.g. see the abstract eval rule for LU) seems to be to just return the operand for `p` as well, even though that's nonsense. I wonder if @hawkinsp, @mattjj, or anyone else can chime in about what semantics we would want for an abstract eval called with non-`ShapedArray` inputs?